src/main/scala/chisel3/experimental/inlinetest/InlineTest.scala

```diff
@@ -88,23 +90,46 @@ object TestHarnessGenerator {
   }
 }
 
+private class TestGenerator[M <: RawModule, R](
+  /** The index of this test in the order they were declared. */
+  val index: Int,
+  /** The user-provided name of the test. */
+  val testName: String,
+  /** The (eventually) legalized name for the DUT module */
+  dutName: () => String,
+  /** The body for this test, returns a result. */
+  testBody: Instance[M] => R,
+  /** The reset type of the DUT module. */
+  resetType: Option[Module.ResetType.Type],
+  /** The testharness generator. */
+  testHarnessGenerator: TestHarnessGenerator[M, R]
+) {
+  def params(dutDefinition: Definition[M]) = new TestParameters(dutName(), testName, dutDefinition, testBody, resetType)
+  def generate(dutDefinition: Definition[M]) = testHarnessGenerator.generate(params(dutDefinition))
+}
+
 /** Provides methods to build unit testharnesses inline after this module is elaborated.
   *
   *  @tparam TestResult the type returned from each test body generator, typically
   *  hardware indicating completion and/or exit code to the testharness.
   */
 trait HasTests[M <: RawModule] { module: M =>
 
-  /** A Definition of the DUT to be used for each of the tests. */
-  private lazy val moduleDefinition =
-    module.toDefinition.asInstanceOf[Definition[module.type]]
+  /** Generators for inline tests by name. */
+  private val testGenerators = new mutable.HashMap[String, TestGenerator[M, _]]
 
-  /** Generate an additional parent around this module.
-    *
-    *  @param parent generator function, should instantiate the [[Definition]]
-    */
-  protected final def elaborateParentModule(parent: Definition[module.type] => RawModule with Public): Unit =
-    afterModuleBuilt { Definition(parent(moduleDefinition)) }
+  lazy val moduleDefinition = module.toDefinition.asInstanceOf[Definition[M]]
```

Comment:
Do we want this to be public? The user might accidentally access it while elaborating the module which would probably cause some havok.